pkg/ratelimiter/ratelimiter.go

```diff
@@ -0,0 +1,124 @@
+package ratelimiter
+
+import (
+	"math"
+	"sync"
+	"time"
+
+	"github.com/argoproj/argo-cd/v2/util/env"
+	"golang.org/x/time/rate"
+	"k8s.io/client-go/util/workqueue"
+)
+
+type AppControllerRateLimiterConfig struct {
+	BucketSize      int
+	BucketQPS       int
+	FailureCoolDown time.Duration
+	BaseDelay       time.Duration
+	MaxDelay        time.Duration
+	BackoffFactor   float64
+}
+
+func GetAppRateLimiterConfig() *AppControllerRateLimiterConfig {
+	return &AppControllerRateLimiterConfig{
+		// global queue rate limit config
+		int(env.ParseInt64FromEnv("WORKQUEUE_BUCKET_SIZE", 500, 1, math.MaxInt64)),
```

Comment:
Should we configure these in an application-controller CLI flag (which falls back to the env var) to improve discoverability?

Reply:
Updated to add cli flags